### ISSUE TITLE
Nim tries to resolve generic field type name even though when condition is false

### ISSUE THREAD

**Issue by endragor** [nim-lang/Nim#4774]
Sample:

``` nim
const hasThreadSupport = compileOption("threads")

type
  SomeObj*[T] = object
    when hasThreadSupport:
      channel: ptr Channel[T]
      # Error: undeclared identifier: 'Channel'
```

The error occurs when you try to compile this without `--threads:on` option.
That is only specific to generic types.

Nim Git hash: 9784e364246d23d4ac8a0fe1466f877594d4a1d0

**Comment by dom96** [comment]
Hrm, could this be the same as #4797?

**Comment by endragor** [comment]
Yeah, looks the same.

**Comment by nc-x** [comment]
The reason is that in generic contexts, we do not set the correct branch to be evaluated and instead evaluate all the branches https://github.com/nim-lang/Nim/blob/ab72d68ec80300ebd0629c5174a78c73f28fc729/compiler/semtypes.nim#L663-L668

It can be fixed currently for simple boolean expressions I think.
But it cannot be fixed currently for expressions like `when T is SomeType`.
Because semIs does not evaluate the expression at that moment so we cannot know which branch to take https://github.com/nim-lang/Nim/blob/ab72d68ec80300ebd0629c5174a78c73f28fc729/compiler/semexprs.nim#L407-L409

IDK what can be done here.

**Comment by Araq** [comment]
This is really nasty to fix...

**Comment by metagn** [comment]
Was fixed in #24042
